### ISSUE TITLE
Enhancement: Convert Profiles from XML to Sqlite

### ISSUE THREAD

**Issue by concubidated** [stepmania/stepmania#1354]
### Summery
Currently the Stats profiles are saved in XML format.  Proposed plan is for profiles to use sqlite databases instead. This will allow better consistency, easier handling of saved data, and simplified profile merging. This issue should be used for discussion and progress tracking.

### Detailed Description

Stepmania will still need to be able to read the XML files in order to maintain previous saved scores. SM should first, check for stats.db, then if there is no stats.db, it should load stats.xml.

_At this point we can init the db, or wait until profile save?_

If waiting for profile save, if the stats.db does not exist, it should init the DB, and save all of the stats.xml profile data to the database. Perhaps renaming, or compressing, the stats.xml for archival purpose should be done here (I don't think we should delete it.)

The MachineProfile Database can remain open until application close. This will allow access to various data without having to have the entire profile loaded in memory. (I believe thats how it currently works). Since sqlite is very good with journaled transactions, failure should not be a large concern here. Writes will be quite limited during normal gameplay. Profile saving should be it. 

Memcard profiles are a bit different. We probably want to do this all in memory and close and save the database to the card at profile save.

_Does this sound reasonable?_

This method appears to be documented well here. http://www.sqlite.org/backup.html

### Current DB schema proposal (WIP)
**TODO**
- Course Data
- Coin Data
![Stepmania DB](http://concubidated.com/smdb.png)

I will finish up the UML tonight before discussing it. Coin Data needs a big reboot and doing it in a DB should solve the shortcomings the current design. _week N for example_

**Comment by dguzek** [comment]
I don't have anything especially useful to add at this point, but I am wondering if _HighScore_ and _HighScoreList_ should both be pluralized to stay consistent with the other tables here.  

(I'm personally more used to table names being pluralized with my Ruby on Rails background, but I don't care what we do here so long as its sensible and consistent.)

**Comment by RGates94** [comment]
HighScore and HighScoreList are both classes that already exist in the code, I don't see a reason for them not to be pluralized, but it would require going through and changing the references to it that already exist.

**Comment by concubidated** [comment]
I was basing the table names on the XML nodes. Plural does make more sense for _HighScore_ and _HighScoreList_. Updated the chart to reflect that.

**Comment by kyzentun** [comment]
You need to have direct support for rate mods in your data model.  Storing all scores in one list and parsing the rate mod out of the modifier string makes per-rate retrieval more difficult and causes problems when scores at one rate bump scores at another out of the limited number of slots.

**Comment by concubidated** [comment]
@kyzentun I agree, in my first diagram I had mods as a separate table all together, but I put it back in  HighScores Table to match the current (not great) design of the XML. Can even then use the mods table to determine if a score should be marked for DQ or not. (Which can/should be theme dependent). The point of this issue is to work out a solid design before implementing, so keep the feedback coming!

**Comment by kyzentun** [comment]
I'm saying rate mods need to be treated differently than any other modifier.  Store one list of scores for each unique rate a chart has been played at.  Make the theme specify what rate it wants when retrieving scores for display.

**Comment by shakesoda** [comment]
If we store the mods we care about in a table to begin with, we can just query for a certain rate anyways.

**Comment by concubidated** [comment]
Added a Modifier Table, basically each mod used will have its own row. This way we can easily do something like this to grab all songs played at a rate of .9

`SELECT * FROM HighScores as hs inner join Modifers as m ON m.HighScoreID = hs.id WHERE m.type="rate" and m.value=".9"`

Obviously we could do this for any mod, which for example, can make it up to the themer to determine what scores should be DQ or not based on the mods used.

**Comment by karai17** [comment]
what about separating this into several DBs so that you have one for the machine (high scores, coin data, etc) and then one for each profile. this would let someone pop their USB stick into an ITG2 cab and keep their own profile data with them.

**Comment by concubidated** [comment]
That was my thought, was each 'stats.xml' would eventually end up with its own DB file.

**Comment by karai17** [comment]
So a few things I'd like to see in user profiles are:

* Record a whole round so that you can always generate a detailed graph or an auto-replay of any given round. This would be useful for adding a ghost graph behind your current play's graph so you can see the areas in a song where you have improved or stumbled, or have a ghost of your best play as P2 for encouragement. I'm not 100% sure on implementation, but it would be nice to have timing info so you can apply any judgement scale to any previous round at any time.
* Session data! I want to know how many steps I've taken during a session, among other stats. Useful when you're using SM for exercise, as I am.

**Comment by shakesoda** [comment]
replays in the db is a *lot* of data to be throwing around, not a good place for them

**Comment by karai17** [comment]
Perhaps only hold on to the top 3 or top 5 or something to that effect. Maybe lock it to the machine's number of high scores per player. Really, you only need your best score to ghost against, but it would be interesting to have as much data as allowable. It could allow a player to visualise their progress over time, for instance.

**Comment by kyzentun** [comment]
Keeping best and worst would be interesting, for players trying to pass a song with the worst possible score. (see various attempts to pass a song on ITG with a score of 0%)

But yeah, replays in the DB would get too big to be wieldy.  If millisecond timing of each press and release is all it is, that means at least 5 bytes per event (1 byte for column and press/release flag, 4 bytes for milliseconds since song start).

That naive approach doesn't work when music rate and especially the haste modifier are introduced.  The current music rate at any given point in the song would also have to be recorded.

**Comment by karai17** [comment]
If not in a database, what other suggestions are there?

**Comment by shakesoda** [comment]
simple compressed binary files with replay data, db can just have the filename

**Comment by karai17** [comment]
Yeah, I dig that.

**Comment by luizsan** [comment]
My idea: add the option to save user/theme defined key-value pairs.

\<MrThatKid\> Or maybe a minigame simfile wants to save the highscore somewhere. Custom fields might work for that?
\<Luizsan\> what I had in mind was saving custom scoring for certain game types *cough*pump
\<Luizsan\> or even multiple scorings
\<Luizsan\> anyway, that would be a nice thing to have 
\<Luizsan\> please consider!
